### PR TITLE
refactor(schema): change the parent class of Fields to MutableMapping

### PR DESCRIPTION
**Pull request by Lee-000**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2840ymr 修改 Fields 父类为 MutableMapping](https://app.clickup.com/t/2840ymr)

File: graviti/utility/user.py [hunk 1]
```diff
@@ -355,3 +357,145 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
 
         """
         self._data.update(__m, **kwargs)
+
+
+class OrderedDict(MutableMapping[str, _V], UserMapping[str, _V]):
```
Comment (linjiX) [inline graviti/utility/user.py:362]: `UserOrderedDict`？

Comment (linjiX) [reply graviti/utility/user.py:362]: If you need to rewrite the `__getitem__` method, you should not set `UserMapping` as base class

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/utility/user.py [hunk 2]
```diff
@@ -355,3 +357,145 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+            self._keys.append(key)
+            self._data[key] = value
+
+    def __getitem__(self, key: Union[int, str]) -> _V:
+        if isinstance(key, int):
+            key = self._keys.__getitem__(key)
+
+        return self._data.__getitem__(key)
+
+    def __setitem__(self, key: str, value: _V) -> None:
```
Comment (linjiX) [inline graviti/utility/user.py:387]: shall we support int key?

File: graviti/utility/user.py [hunk 3]
```diff
@@ -355,3 +357,145 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+        if not self._data.__contains__(key):
+            self._keys.append(key)
+
+        self._data.__setitem__(key, value)
+
+    def __delitem__(self, key: Union[int, str]) -> None:
+        if isinstance(key, int):
+            self._keys.__delitem__(key)
+            key = self._keys.__getitem__(key)
```
Comment (linjiX) [inline graviti/utility/user.py:396]: get after delete?
why not try list.pop?

File: graviti/utility/user.py [hunk 4]
```diff
@@ -355,3 +357,145 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+        return self._data.__contains__(key)
+
+    def __iter__(self) -> Iterator[str]:
+        return self._keys.__iter__()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            return False
+
+        return self._data.__eq__(other._data) and self._keys.__eq__(other._keys)
```
Comment (linjiX) [inline graviti/utility/user.py:412]: shall we check `keys` first?
will it be faster?

File: graviti/utility/user.py [hunk 5]
```diff
@@ -355,3 +357,145 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+    def keys(self) -> Tuple[str, ...]:  # type: ignore[override]
+        """Get all element keys in order.
+
+        Returns:
+            A tuple containing all elements keys.
+
+        """
+        return tuple(self._keys)
+
+    def items(self) -> Tuple[Tuple[str, _V], ...]:  # type: ignore[override]
```
Comment (linjiX) [inline graviti/utility/user.py:423]: shall we return an `Iterator`?

File: graviti/utility/user.py [hunk 6]
```diff
@@ -355,3 +357,145 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+
+    @overload
+    def pop(self, key: str) -> _V:  # pylint: disable=arguments-differ
+        ...
+
+    @overload
+    def pop(self, key: str, default: Union[_V, _T] = ...) -> Union[_V, _T]:
+        ...
+
+    def pop(self, key: str, default: Any = __marker) -> Any:
```
Comment (linjiX) [inline graviti/utility/user.py:454]: shall we support int key?

File: graviti/utility/user.py [hunk 7]
```diff
@@ -355,3 +357,145 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+        Pairs are returned in LIFO (last-in, first-out) order.
+
+        Returns:
+            A (key, value) pair as a tuple.
+
+        """
+        key = self._keys.pop()
+        return (key, self._data.pop(key))
+
+    def setdefault(self, key: str, default: _V = None) -> _V:  # type: ignore[assignment]
```
Comment (linjiX) [inline graviti/utility/user.py:485]: int key?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/utility/user.py (hunk 5 above)
Comment (Lee-000) [reply graviti/utility/user.py:423]: If `items()` returns `Iterator`, should `keys()` and `values()` also return `Iterator`?
In the current situation, there wouldn't be too many items in it.

File: graviti/utility/user.py (hunk 7 above)
Comment (Lee-000) [reply graviti/utility/user.py:485]: I think it is wired to support int key here. `__setitem__ should be used to set the value at a specific index.

File: graviti/utility/user.py (hunk 5 above)
Comment (linjiX) [reply graviti/utility/user.py:433]: obviously Iterator is better than tuple

File: graviti/utility/user.py [hunk 8]
```diff
@@ -355,3 +357,178 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
 
         """
         self._data.update(__m, **kwargs)
+
+
+class UserOrderedDict(MutableMapping[str, _V]):
```
Comment (linjiX) [inline graviti/utility/user.py:362]: can you get a better name?

File: graviti/utility/user.py [hunk 9]
```diff
@@ -355,3 +357,178 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+
+    @overload
+    def get(self, key: str) -> Optional[_V]:  # pylint: disable=arguments-differ
+        ...
+
+    @overload
+    def get(self, key: str, default: Union[_V, _T] = ...) -> Union[_V, _T]:
+        ...
+
+    def get(self, key: str, default: Any = None) -> Any:
```
Comment (linjiX) [inline graviti/utility/user.py:464]: support int key?

File: graviti/utility/user.py [hunk 10]
```diff
@@ -355,3 +357,178 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+        if not self._data.__contains__(key):
+            self._keys.append(key)
+
+        self._data.__setitem__(key, value)
+
+    def __delitem__(self, key: Union[int, str]) -> None:
+        if isinstance(key, int):
+            key = self._keys.pop(key)
+        else:
+            self._keys.remove(key)
```
Comment (linjiX) [inline graviti/utility/user.py:408]: what if the str key is not in `keys`?
what exception should we raise?

File: graviti/utility/user.py (hunk 7 above)
Comment (linjiX) [reply graviti/utility/user.py:518]: why it is wired?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/builtin.py [hunk 11]
```diff
@@ -261,29 +261,14 @@ class record(PortexBuiltinType):  # pylint: disable=invalid-name
           },
         )
 
-        Create a record by :class:`Field` list:
-
-        >>> t = record([Field("f0", float64(0)), Field("f1", array(int64()))])
```
Comment (linjiX) [inline graviti/schema/builtin.py:266]: Field is also tuple, it is legal input of OrderDict, shall we put the Field remove job into another PR?

File: graviti/utility/user.py [hunk 12]
```diff
@@ -355,3 +357,178 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+        Arguments:
+            key: The key for dict.
+            default: The value to be returned if the key is not in the dict and it is given.
+
+        Returns:
+            Value to be removed from the UserOrderedDict object.
+
+        """
+        if isinstance(key, int):
+            key = self._keys.pop(key)
```
Comment (linjiX) [inline graviti/utility/user.py:497]: what if the `self._keys.pop` raise `IndexError`?
if the default is given, `pop` should not raise when the key does not exist.

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/utility/user.py [hunk 13]
```diff
@@ -355,3 +357,178 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+        elif self._data.__contains__(key):
+            self._keys.remove(key)
+
+        if default is self.__marker:
+            return self._data.pop(key)
+
+        return self._data.pop(key, default)
+
+    def popitem(self) -> Tuple[str, _V]:
```
Comment (linjiX) [inline graviti/utility/user.py:506]: what exception should `popitem` raise when the dict is empty?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/utility/user.py (hunk 9 above)
Comment (Lee-000) [reply graviti/utility/user.py:464]: We only support int in those methods that exit in `list` as well currently.

File: graviti/utility/user.py [hunk 14]
```diff
@@ -355,3 +357,86 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+    def __iter__(self) -> Iterator[str]:
+        return self._keys.__iter__()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            return False
+
+        return self._keys.__eq__(other._keys) and self._data.__eq__(other._data)
+
+    def popitem(self) -> Tuple[str, _V]:
```
Comment (linjiX) [inline graviti/utility/user.py:425]: why we need `popitem` here?
basicly `popitem` is a remove operation, but this `popitem` has not called `__delitem__`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/utility/user.py [hunk 15]
```diff
@@ -355,3 +357,86 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
 
         """
         self._data.update(__m, **kwargs)
+
+
+class StringKeyOrderedDict(MutableMapping[str, _V]):
```
Comment (linjiX) [inline graviti/utility/user.py:362]: this name is too long, can we get a better name?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/utility/user.py (hunk 15 above)
Comment (linjiX) [reply graviti/utility/user.py:362]: also shall we still put this class in `user.py`?

File: graviti/utility/user.py (hunk 14 above)
Comment (Lee-000) [reply graviti/utility/user.py:425]: Because if use the `MutableMapping.popitem` method, the first item will be popped. And we want it to popitem from the tail.

Comment (linjiX) [reply graviti/utility/user.py:425]: Ok, we need to call `self.__delitem__` in popitem

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/utility/user.py [hunk 16]
```diff
@@ -355,3 +357,88 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
 
         """
         self._data.update(__m, **kwargs)
+
+
+class SimpleOrderedDict(MutableMapping[str, _V]):
```
Comment (linjiX) [inline graviti/utility/user.py:362]: This name is too simple, can you get a better name?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/utility/user.py [hunk 17]
```diff
@@ -355,3 +357,88 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+        Returns:
+            A (key, value) pair as a tuple.
+
+        """
+        try:
+            key = self._keys[-1]
+        except IndexError:
+            raise KeyError("popitem(): dictionary is empty") from None
+
+        value = self.__getitem__(key)
```
Comment (linjiX) [inline graviti/utility/user.py:442]: can we use `self.pop`, `pop` calls `__delitem__`

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/utility/user.py [hunk 18]
```diff
@@ -355,3 +357,87 @@ def update(self, __m: Any = (), **kwargs: _V) -> None:  # pylint: disable=argume
...
+        Returns:
+            A (key, value) pair as a tuple.
+
+        """
+        try:
+            key = self._keys[-1]
+        except IndexError:
+            raise KeyError("popitem(): dictionary is empty") from None
+
+        value = self.pop(key)
```
Comment (linjiX) [inline graviti/utility/user.py:442]: `return key, self.pop(key)`

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — APPROVED** [review]
